### ISSUE TITLE
Best practices for interfacing tf.Transform and tf.data.Dataset

### ISSUE THREAD

**Issue by alshedivat** [tensorflow/transform#83]
I was looking for the best way to replicate the training data preprocessing at serving time. It looks like `tf.Transform` is the way to go, but it's unclear, what are the best practice of interfacing it with `tf.data.Dataset` pipelines (which also have dataset mappers, etc.)?

Also, going forward, will `tf.Transform` and `tf.data.Dataset` co-exist, or one will succeed the other?

**Comment by harshini-gadige** [comment]
@zoyahav  - Hi, could you please look into this issue ?

**Comment by NikeNano** [comment]
Hi, 

You could check out how they did in this example from the google developers blog:
https://github.com/Fematich/tftransform-demo/blob/master/trainer/util.py. 

Blog post here:https://cloud.google.com/blog/products/ai-machine-learning/pre-processing-tensorflow-pipelines-tftransform-google-cloud

**Comment by hgadig** [comment]
@alshedivat  Did you get a chance to go through the above references ? Please keep us posted.

**Comment by alshedivat** [comment]
@NikeNano, @hgadig, thanks for pointing out to the resources. They seem to be quite relevant, but not exactly what I was looking for.

In the `utils.py` file, input and serving functions already assume that the data is pooled from some transformed TFRecords. In the blog, again, transformed data is saved before being used later on.

I wonder `tf.Transform` can be used in streaming scenarios. E.g., my typical `tf.data.Dataset` pipeline for image processing pools raw images from the disk, then extracts regions of interst, applies normalization and/or other preprocessing, applies perturbations on the fly (if necessary), and finally forms inputs out of that. I think if this as working with an infinite stream of data (at training time). At serving time, I need to make sure my training transformations are correctly replicated (ideally, I'd like to be able to somehow "checkpoint" transformations along with the best model I get from training).

Thanks again for your help!

**Comment by KesterTong** [comment]
@alshedivat You can do everything except applying the perturbations with Transform.  I would advise applying peturbations with tf.Dataset.  All our examples use both Transform and tf.dataset together.

The only thing that is not supported is streaming.  As you say, Transform assumes batch computation.  But it wasn't clear to me if streaming was a requirement or just how you currently do your processing.

**Comment by hgadig** [comment]
Closing this as it is in "awaiting response" for more than a week. Feel free to add your comments(if any) and we will reopen. Thanks!
